### ISSUE TITLE
[inductor][perf] Inductor/Triton softmax kernel is slower than eager

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#129104]
```py
import torch
import time


def eager(a, b):
    return torch.softmax(a, -1)


@torch.compile(options={"cuda_backend": "triton", "cpu_backend": "cpp"})
def triton(a, b):
    return eager(a, b)


def benchmark(fn, shape=(8192 * 4, 8192), niter=100, warmup=10, repeat=10):
    a = torch.rand(shape, device="cuda")
    b = torch.rand(shape, device="cuda")
    sync = torch.cuda.synchronize

    for _ in range(warmup):
        result = fn(a, b)
        torch.testing.assert_close(result, eager(a, b))

    results = []
    for _ in range(repeat):
        sync()
        start = time.perf_counter()
        for _ in range(niter):
            fn(a, b)
        sync()
        end = time.perf_counter()
        results.append((end - start) * 1000 / niter)

    # min to smooth out any noise from context switches
    print(f"{fn.__name__} took {min(results):.4f} ms")


benchmark(eager)
benchmark(triton)
```

Output on my local RTX 3090:
```
python softmax.py 
eager took 2.6877 ms
triton took 4.3575 ms
```

The generated Triton code:
```
def triton_(in_ptr0, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 32768
    rnumel = 8192
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = tl.full([XBLOCK, RBLOCK], True, tl.int1)
    rbase = tl.arange(0, RBLOCK)[None, :]
    x0 = xindex
    _tmp2 = tl.full([XBLOCK, RBLOCK], float("-inf"), tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
        tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
        tmp3 = triton_helpers.maximum(_tmp2, tmp1)
        _tmp2 = tl.where(rmask, tmp3, _tmp2)
    tmp2 = triton_helpers.max2(_tmp2, 1)[:, None]
    _tmp8 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp4 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
        tmp5 = tmp4 - tmp2
        tmp6 = tl_math.exp(tmp5)
        tmp7 = tl.broadcast_to(tmp6, [XBLOCK, RBLOCK])
        tmp9 = _tmp8 + tmp7
        _tmp8 = tl.where(rmask, tmp9, _tmp8)
    tmp8 = tl.sum(_tmp8, 1)[:, None]
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp10 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_first', other=0.0)
        tmp11 = tmp10 - tmp2
        tmp12 = tl_math.exp(tmp11)
        tmp13 = tmp12 / tmp8
        tl.store(out_ptr2 + (r1 + (8192*x0)), tmp13, rmask)
```

@eellison @shunting314 @bertmaher @htyu not sure if this is a case where inductor's generate code needs to be improved, or perhaps it is an issue with Triton -- but in either case it points to some performance gaps in what we are doing.

cc @ezyang @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @bertmaher @int3 @davidberard98 @nmacchioni @embg

**Comment by eellison** [comment]
If we run with either `TORCHINDUCTOR_COORDINATE_DESCENT_TUNING` or force persistent reductions to occur it beats eager. Funny enough the coordinate descent tuning it discovers is essentially a persistent reduction because it makes the `rblock` equal to the length of the loop:

```
:XBLOCK: 1, RBLOCK: 8192, num_warps: 16, num_ctas: 1, num_stages: 1
```

If we force the rblock to be equal to the length of the loop is there any difference between persistent reduction and normal codegen 🤔  assuming that triton can dedup loads. 

I can look into improving the heuristics for this case.

**Comment by shunting314** [comment]
If rnumel is even larger, we would not be able to use persistent reduction.

I think this online softmax PR (https://github.com/pytorch/pytorch/pull/127011 ) would help here.

**Comment by jansel** [comment]
Interesting, perhaps some more evidence of suboptimal heuristics.  cc @AlnisM @Chillee

**Comment by Chillee** [comment]
I think this is pretty similar to some of the experiments I did previously on softmax perf - block size tuning is the main gap here.
